File: Platform.Collections.Methods.Tests/SizeBalancedTree.cs
using System;
using System.Collections.Generic;
using System.Text;
using Platform.Numbers;
using Platform.Collections.Methods.Trees;

namespace Platform.Collections.Methods.Tests
{
    internal class SizeBalancedTree<TElement> : SizeBalancedTreeMethods<TElement>
    {
        private struct TreeElement
        {
            public TElement Size;
            public TElement Left;
            public TElement Right;
            public sbyte Balance;
            public bool LeftIsChild;
            public bool RightIsChild;
        }

        private readonly TreeElement[] _elements;
        private TElement _allocated;

        public TElement Root;

        public TElement Count => GetSizeOrZero(Root);

        public SizeBalancedTree(int capacity) => (_elements, _allocated) = (new TreeElement[capacity], Integer<TElement>.One);

        public TElement Allocate()
        {
            var newNode = _allocated;
            if (IsEmpty(newNode))
            {
                _allocated = Arithmetic.Increment(_allocated);
                return newNode;
            }
            else
            {
                throw new InvalidOperationException("Allocated tree element is not empty.");
            }
        }

        public void Free(TElement node)
        {
            while (!EqualityComparer.Equals(_allocated, Integer<TElement>.One) && IsEmpty(node))
            {
                var lastNode = Arithmetic.Decrement(_allocated);
                if (EqualityComparer.Equals(lastNode, node))
                {
                    _allocated = lastNode;
                    node = Arithmetic.Decrement(node);
                }
                else
                {
                    return;
                }
            }
        }

        public bool IsEmpty(TElement node) => EqualityComparer<TreeElement>.Default.Equals(GetElement(node), default);

        protected override bool FirstIsToTheLeftOfSecond(TElement first, TElement second) => Comparer.Compare(first, second) < 0;

        protected override bool FirstIsToTheRightOfSecond(TElement first, TElement second) => Comparer.Compare(first, second) > 0;

        protected override ref TElement GetLeftReference(TElement node) => ref GetElement(node).Left;

        protected override TElement GetLeft(TElement node) => GetElement(node).Left;

        protected override ref TElement GetRightReference(TElement node) => ref GetElement(node).Right;

        protected override TElement GetRight(TElement node) => GetElement(node).Right;

        protected override TElement GetSize(TElement node) => GetElement(node).Size;

        protected override void PrintNodeValue(TElement node, StringBuilder sb) => sb.Append(node);

        protected override void SetLeft(TElement node, TElement left) => GetElement(node).Left = left;

        protected override void SetRight(TElement node, TElement right) => GetElement(node).Right = right;

        protected override void SetSize(TElement node, TElement size) => GetElement(node).Size = size;

        private ref TreeElement GetElement(TElement node) => ref _elements[(Integer<TElement>)node];
    }
}

File: cpp/Platform.Collections.Methods.Tests/SizeBalancedTree.cpp
namespace Platform::Collections::Methods::Tests
{
    template <std::size_t N, typename ...> class SizeBalancedTree;
    template <std::size_t N, typename TElement> class SizeBalancedTree<N, TElement> : public Platform::Collections::Methods::Trees::SizeBalancedTreeMethods<TElement>
    {
        struct TreeElement
        {
            public: TElement Size = 0;
            public: TElement Left = 0;
            public: TElement Right = 0;
        };

        private: TreeElement _elements[N] = { {0} };
        private: TElement _allocated = 0;

        public: TElement Root = 0;

        public: TElement GetCount() { return this->GetSizeOrZero(Root); }

        public: SizeBalancedTree() { _allocated = 1; }

        public: TElement Allocate()
        {
            auto newNode = _allocated;
            if (this->IsEmpty(newNode))
            {
                _allocated = _allocated + 1;
                return newNode;
            }
            else
            {
                throw std::runtime_error("Allocated tree element is not empty.");
            }
        }

        public: void Free(TElement node)
        {
            while (_allocated != 1 && this->IsEmpty(node))
            {
                auto lastNode = _allocated - 1;
                if (lastNode == node)
                {
                    _allocated = lastNode;
                    node = node - 1;
                }
                else
                {
                    return;
                }
            }
        }

        public: bool IsEmpty(TElement node) { return iszero(this->GetElement(node), sizeof(TreeElement)); }

        protected: bool FirstIsToTheLeftOfSecond(TElement first, TElement second) override { return first < second; }

        protected: bool FirstIsToTheRightOfSecond(TElement first, TElement second) override { return first > second; }

        protected: TElement* GetLeftReference(TElement node) override { return &GetElement(node)->Left; }

        protected: TElement GetLeft(TElement node) override { return this->GetElement(node)->Left; }

        protected: TElement* GetRightReference(TElement node) override { return &GetElement(node)->Right; }

        protected: TElement GetRight(TElement node) override { return this->GetElement(node)->Right; }

        protected: TElement GetSize(TElement node) override { return this->GetElement(node)->Size; }

        protected: void SetLeft(TElement node, TElement left) override { this->GetElement(node)->Left = left; }

        protected: void SetRight(TElement node, TElement right) override { this->GetElement(node)->Right = right; }

        protected: void SetSize(TElement node, TElement size) override { this->GetElement(node)->Size = size; }

        private: TreeElement* GetElement(TElement node) { return &_elements[node]; }
    };
}
